disco_aws_automation/disco_vpc.py

```diff
@@ -641,307 +429,63 @@ def _destroy_rds(self, wait=True):
-
-    def get_all_security_groups_for_vpc(self):
-        """ Find all security groups belonging to vpc and return them """
-        return self.vpc.connection.get_all_security_groups(filters=self.vpc_filter())
-
-    def _destroy_igws(self):
-        """ Find all gateways belonging to vpc and destroy them"""
-        vpc_attachment_filter = {"attachment.vpc-id": self.vpc.id}
-        # delete gateways
-        for igw in self.vpc.connection.get_all_internet_gateways(filters=vpc_attachment_filter):
-            self.vpc.connection.detach_internet_gateway(igw.id, self.vpc.id)
-            self.vpc.connection.delete_internet_gateway(igw.id)
+        for subnet in self.boto3_ec2.describe_subnets(Filters=self.vpc_filters())['Subnets']:
+            self.boto3_ec2.delete_subnet(SubnetId=subnet['SubnetId'])
 
     def _destroy_routes(self):
         """ Find all route_tables belonging to vpc and destroy them"""
-        for route_table in self.vpc.connection.get_all_route_tables(filters=self.vpc_filter()):
-            if not route_table.tags:
-                logging.info("Skipping untagged (default) route table %s", route_table.id)
+        for route_table in self.boto3_ec2.describe_route_tables(Filters=self.vpc_filters())['RouteTables']:
+            if len(route_table["Associations"]) > 0 and route_table["Associations"][0]["Main"]:
+                logging.info("Skipping the default main route table %s", route_table['RouteTableId'])
                 continue
             try:
-                self.vpc.connection.delete_route_table(route_table.id)
+                self.boto3_ec2.delete_route_table(RouteTableId=route_table['RouteTableId'])
             except EC2ResponseError:
-                logging.error("Error deleting route_table %s:.", route_table.id)
+                logging.error("Error deleting route_table %s:.", route_table['RouteTableId'])
                 raise
 
     def _destroy_vpc(self):
         """Delete VPC and then delete the dhcp_options that were associated with it. """
 
-        # save function and parameters so we can delete dhcp_options after vpc. We do this becase botos
-        # get_all_dhcp_options does not support filter. Because we cannot easily find the default dhcp
-        # options, re-assigning default dhcp option is not trivial.
-        delete_dhcp_options = self.vpc.connection.delete_dhcp_options
-        dhcp_options_id = self.vpc.dhcp_options_id
+        # save function and parameters so we can delete dhcp_options after vpc.
+        dhcp_options_id = self.vpc['DhcpOptionsId']
 
-        delete_status = keep_trying(30, self.vpc.delete)
+        self.boto3_ec2.delete_vpc(VpcId=self.get_vpc_id())
         self.vpc = None
 
-        if not delete_dhcp_options(dhcp_options_id):
-            logging.warning("failed to delete dhcp options (%s)", dhcp_options_id)
-
-        return delete_status
+        self.boto3_ec2.delete_dhcp_options(DhcpOptionsId=dhcp_options_id)
```

Comment:
We used to accept failure of deleting dhcp now we don't. I hit following error:

```
(asiaq) isukhanov@isukhanov-dskt ~/v/asiaq_astro_config (master) [1]> disco_vpc_ui.py destroy --name ias
2016-06-03 14:42:57,478 root INFO Skipping the default main route table rtb-38add05c
Traceback (most recent call last):
  File "/home/isukhanov/.virtualenvs/asiaq/bin/disco_vpc_ui.py", line 10, in <module>
    execfile(__file__)
  File "/home/isukhanov/vcs/asiaq/bin/disco_vpc_ui.py", line 184, in <module>
    run_gracefully(run)
  File "/home/isukhanov/vcs/asiaq/disco_aws_automation/disco_aws_util.py", line 61, in run_gracefully
    main_function()
  File "/home/isukhanov/vcs/asiaq/bin/disco_vpc_ui.py", line 174, in run
    destroy_vpc_command(args)
  File "/home/isukhanov/vcs/asiaq/bin/disco_vpc_ui.py", line 98, in destroy_vpc_command
    vpc.destroy()
  File "/home/isukhanov/vcs/asiaq/disco_aws_automation/disco_vpc.py", line 410, in destroy
    self._destroy_vpc()
  File "/home/isukhanov/vcs/asiaq/disco_aws_automation/disco_vpc.py", line 475, in _destroy_vpc
    self.boto3_ec2.delete_dhcp_options(DhcpOptionsId=dhcp_options_id)
  File "/home/isukhanov/.virtualenvs/asiaq/local/lib/python2.7/site-packages/botocore/client.py", line 258, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/isukhanov/.virtualenvs/asiaq/local/lib/python2.7/site-packages/botocore/client.py", line 548, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (DependencyViolation) when calling the DeleteDhcpOptions operation: The dhcpOptions 'dopt-c52ad2a0' has dependencies and cannot be deleted.
```

Maybe VPC deletion was not complete when this was called?